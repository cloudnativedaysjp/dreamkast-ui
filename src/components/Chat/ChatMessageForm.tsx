import React, { useEffect, useState } from 'react'
import { useForm } from 'react-hook-form'
import {
  ChatMessageApi,
  ChatMessageMessageTypeEnum,
  Configuration,
} from '../../client-axios'
import Button from '@material-ui/core/Button'
import { ChatMessageClass } from './index'

type Props = {
  roomId?: number
  selectedMessage: ChatMessageClass
  onClickCloseButton: () => void
}

type Inputs = {
  chatMessage: string
  isQuestion: boolean
}

const ChatMessageRequest = (
  eventAbbr: string,
  roomId: number,
  roomType: string,
  body: string,
  replyTo: number | null,
) => {
  return {
    eventAbbr: eventAbbr,
    roomId: roomId,
    roomType: roomType,
    body: body,
    messageType: ChatMessageMessageTypeEnum.Chat,
    replyTo: replyTo,
  }
}

const ChatMessageForm: React.FC<Props> = ({
  roomId,
  selectedMessage,
  onClickCloseButton,
}) => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { isSubmitSuccessful },
  } = useForm<Inputs>()
  const [submittedData, setSubmittedData] = useState({})
  const messageSelected = !!selectedMessage.id
  const createChatMessageRequest = (data: Inputs, roomId: number) => {
    const req = ChatMessageRequest(
      'cndo2021',
      roomId,
      'talk',
      data.chatMessage,
      null,
    )
    if (selectedMessage && selectedMessage.id) {
      req.replyTo = selectedMessage.id
    }
    if (data.isQuestion) {
      req.messageType = ChatMessageMessageTypeEnum.Qa
    }
    return req
  }
  useEffect(() => {
    console.log(selectedMessage)
    console.log(messageSelected)
  }, [])

  const onSubmit = (data: Inputs) => {
    if (!roomId) return
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )
    setSubmittedData(data)
    api.apiV1ChatMessagesPost(createChatMessageRequest(data, roomId))
  }

  const onThumbsUp = () => {
    if (!roomId) return
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )

    const msg = ChatMessageRequest('cndo2021', roomId, 'talk', '👍', null)
    api.apiV1ChatMessagesPost(msg)
  }
  const onClap = () => {
    if (!roomId) return
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )

    const msg = ChatMessageRequest('cndo2021', roomId, 'talk', '👏', null)
    api.apiV1ChatMessagesPost(msg)
  }
  const onPartyPopper = () => {
    if (!roomId) return
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )

    const msg = ChatMessageRequest('cndo2021', roomId, 'talk', '🎉', null)
    api.apiV1ChatMessagesPost(msg)
  }

  useEffect(() => {
    if (isSubmitSuccessful) {
      reset({ chatMessage: '' })
    }
  }, [isSubmitSuccessful, submittedData, reset])

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {messageSelected && <p onClick={onClickCloseButton}>Close</p>}
      {messageSelected && <p>{selectedMessage.body}</p>}
      <textarea name="chatMessage" ref={register} />
      <input type="submit" />
      <br />
      {!messageSelected && (
        <div>
          <input type="checkbox" name="isQuestion" ref={register} />
          質問を送る
          <br />
        </div>
      )}
      <Button onClick={onThumbsUp} variant="contained">
        👍
      </Button>
      <Button onClick={onClap} variant="contained">
        👏
      </Button>
      <Button onClick={onPartyPopper} variant="contained">
        🎉
      </Button>
    </form>
  )
}

export default ChatMessageForm
