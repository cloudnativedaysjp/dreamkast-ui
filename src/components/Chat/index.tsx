export { Chat } from './Chat'
import { ChatMessage as ChatMessageInterface, ChatMessageMessageTypeEnum } from '../../client-axios'


export class ChatMessageClass implements ChatMessageInterface {
  id?: number;
  profileId? :number;
  speakerId?: number;
  eventAbbr: string;
  roomId?: number;
  roomType?: string;
  body: string;
  messageType: ChatMessageMessageTypeEnum;

  constructor(id: number, profileId: number, speakerId: number, eventAbbr: string, roomId: number, roomType: string, body: string, messageType: ChatMessageMessageTypeEnum) {
    this.id = id;
    this.profileId = profileId;
    this.speakerId = speakerId;
    this.eventAbbr = eventAbbr;
    this.roomId = roomId;
    this.roomType = roomType;
    this.body = body;
    this.messageType = messageType
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  createFromWebSocket(
    id: number,
    eventAbbr: string,
    roomId: number,
    roomType: string,
    body: string,
  ) {
    this.id = id
    this.eventAbbr = eventAbbr
    this.roomId = roomId
    this.roomType = roomType
    this.body = body
  }
}
