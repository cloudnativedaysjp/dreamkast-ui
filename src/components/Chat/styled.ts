import styled from 'styled-components'
import { Paper, Box as MaterialUIBox } from '@material-ui/core'

export const Box = styled(MaterialUIBox)`
  height: 400px;
`

export const ChatMessage = styled(Paper)<{ isChat: boolean }>`
  padding: 10px;
  background: ${({ isChat }) => (isChat ? '#ffffff' : '#f5f5f5')};
  white-sace: pre-wrap;
`
export const ChatReplyMessage = styled(ChatMessage)`
  margin-left: 20px;
`

export const ChatSelectedMessage = styled(ChatMessage)`
  background-color: cornsilk;
`

export const CloseReplyButton = styled.div`
  cursor: pointer;
  float: right;
  margintop: 5px;
  width: 20px;
`
