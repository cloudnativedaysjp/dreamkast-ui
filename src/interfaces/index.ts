// You can include shared interfaces/types in a separate file
// and then use them in any component by importing them. For
// example, to import the interface below do:
//
// import { User } from 'path/to/interfaces';

import {ChatMessage as ChatMessageInterface} from "../client-axios";

export type Talk = {
  id: string
  trackId: string
  vimeoId: string
  title: string
  description: string
  speakers: string[]
  onAir?: boolean
}

export type Track = {
  [key: string]: string
}

export class ChatMessageClass implements ChatMessageInterface {
  id?: number
  eventAbbr: string;
  roomId?: number;
  roomType?: string
  body: string;

  constructor(eventAbbr: string, roomId: number, roomType: string, body: string) {
    this.eventAbbr = eventAbbr;
    this.roomId = roomId;
    this.roomType = roomType;
    this.body = body;
  }

  createFromWebSocket(id: number, eventAbbr: string, roomId: number, roomType: string, body: string) {
    this.id = id
    this.eventAbbr = eventAbbr;
    this.roomId = roomId;
    this.roomType = roomType;
    this.body = body;
  }
}
