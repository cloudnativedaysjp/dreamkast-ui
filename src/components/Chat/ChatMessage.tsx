import React, { useEffect, useRef } from 'react'
import * as Styled from './styled'
import {
  ChatMessage as ChatMessageObject,
  ChatMessageMessageTypeEnum,
} from '../../client-axios/api'

type Props = {
  chatMessage: ChatMessageObject
  setRef: (
    chatMessage: ChatMessageObject,
    ref: React.RefObject<HTMLDivElement>,
  ) => void
}

const ChatMessage: React.FC<Props> = ({ chatMessage, setRef }) => {
  const ref = useRef<HTMLDivElement>(null)
  const isSpeakerMessage = !!chatMessage.speaker_id

  useEffect(() => {
    setRef(chatMessage, ref)
  }, [setRef, chatMessage, ref])

  return (
    <div ref={ref}>
      <Styled.ChatMessage
        isChat={chatMessage.messageType == ChatMessageMessageTypeEnum.Chat}
      >
        {isSpeakerMessage ? '[S] ' : ''}
        {chatMessage.body}
      </Styled.ChatMessage>
    </div>
  )
}

export default ChatMessage
