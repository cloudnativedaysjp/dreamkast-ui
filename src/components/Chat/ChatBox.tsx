import React from 'react'
import { Box} from '@material-ui/core'
import { ChatMessage as ChatMessageInterface, ChatMessageMessageTypeEnum } from '../../client-axios'
import {makeStyles} from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import ChatMessage from "./ChatMessage";

type Props = {
  messages: ChatMessageInterface[]
  messageTypes: ChatMessageMessageTypeEnum[]
}
const useStyles = makeStyles((theme) => ({
    box: {
        height: '400px',
    },
    paper: {
        padding: theme.spacing(2),
        color: theme.palette.text.secondary,
    },
}))

export const ChatBox: React.FC<Props> = ({ messages, messageTypes }) => {
  const classes = useStyles()
  const setLastMessageElement = (
    chatMessage: ChatMessageInterface,
    ref: React.RefObject<HTMLDivElement>,
  ) => {
    const lastChat = messages[messages.length - 1]
    if (chatMessage.id === lastChat.id) {
      ref && ref.current && ref.current.scrollIntoView()
    }
  }

  return (
    <Box component="div" overflow="scroll" className={classes.box}>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          {messages.map((chatMessage) => {
            if (messageTypes.includes(chatMessage.messageType)) {
              return (
                <ChatMessage
                  chatMessage={chatMessage}
                  setRef={setLastMessageElement}
                />
              )
            }
          })}
        </Grid>
      </Grid>
    </Box>
  )
}
