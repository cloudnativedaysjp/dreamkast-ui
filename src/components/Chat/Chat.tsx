import React, { useEffect, useState } from 'react'
import * as Styled from './styled'
import {
  Event,
  ChatMessageApi,
  ChatMessageMessageTypeEnum,
  Talk,
  Configuration,
  Profile,
} from '../../client-axios'
import { ChatMessageForm } from './internal/ChatMessageForm'
import ActionCable from 'actioncable'
import dayjs from 'dayjs'
import { ChatMessageClass, ChatMessageMap } from '../../util/chat'
import { TabContext } from '@material-ui/lab'
import { ChatBox } from './internal/ChatBox'
import {
  CreateChatMessageRequest,
  MessageInputs,
} from './internal/ChatMessageRequest'

type Props = {
  event: Event
  profile?: Profile
  talk?: Talk
}

type ReceivedMsg = {
  id: number
  profileId: number
  speakerId: number
  eventAbbr: string
  roomId: number
  roomType: string
  createdAt: string
  body: string
  messageType: ChatMessageMessageTypeEnum
  replyTo: number
}

export const Chat: React.FC<Props> = ({ event, profile, talk }) => {
  const initialChatMessageMap = new ChatMessageMap()
  const initialChatMessage = {
    eventAbbr: event.abbr,
    body: '',
    roomId: !!talk ? talk.id : 0,
    messageType: ChatMessageMessageTypeEnum.Chat,
  }
  const [selectedTab, setSelectedTab] = useState('0')
  const [messages, setMessages] = useState<ChatMessageMap>(
    initialChatMessageMap,
  )
  const [selectedMessage, setSelectedMessage] = useState<ChatMessageClass>(
    initialChatMessage,
  )
  const [chatCable, setChatCable] = useState<ActionCable.Cable | null>(null)
  const [checked, setChecked] = useState<boolean>(true)
  const [isVisibleForm, setIsVisibleForm] = useState<boolean>(true)

  const actionCableUrl = () => {
    if (window.location.protocol == 'http:') {
      return `ws://${window.location.host}/cable`
    } else {
      return `wss://${window.location.host}/cable`
    }
  }
  const fetchChatMessagesFromAPI = async () => {
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )
    const { data } = await api.apiV1ChatMessagesGet(
      event.abbr,
      String(talk?.id),
      'talk',
    )
    if (typeof data !== 'object') {
      // Chatのwebsocketがエラーを返した場合はログインさせるためにトップページへリダイレクト
      window.location.href = `${window.location.href.replace(/\/ui\//g, '')}`
    }
    if (!messages) setMessages(new ChatMessageMap())
    messages.clear()
    data.forEach((receivedMsg) => {
      messages.addMessage(receivedMsg as ChatMessageClass)
    })
    setMessages(new ChatMessageMap(messages))
  }

  const cableReceived = (receivedMsg: ReceivedMsg) => {
    if (!messages) return
    const msg = new ChatMessageClass(
      receivedMsg.id,
      receivedMsg.profileId,
      receivedMsg.speakerId,
      receivedMsg.eventAbbr,
      receivedMsg.roomId,
      receivedMsg.roomType,
      receivedMsg.createdAt,
      receivedMsg.body,
      receivedMsg.messageType,
      receivedMsg.replyTo,
    )
    messages.addMessage(msg)
    setMessages(new ChatMessageMap(messages))
  }

  useEffect(() => {
    if (!talk || !messages) return
    if (chatCable) chatCable.disconnect()

    const endTime = `${talk?.conferenceDayDate} ${dayjs(talk.endTime).format(
      'HH:mm',
    )}`
    // 発表時間の幅を考慮して10分(600秒)余裕をもたせる
    setIsVisibleForm(dayjs().unix() - dayjs(endTime).unix() < 600)

    setSelectedMessage(initialChatMessage)
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const actionCable = require('actioncable')
    fetchChatMessagesFromAPI()
    const wsUrl = actionCableUrl()
    const cable = actionCable.createConsumer(wsUrl)
    setChatCable(cable)
    cable.subscriptions.create(
      { channel: 'ChatChannel', roomType: 'talk', roomId: talk.id },
      { received: cableReceived },
    )
  }, [talk])

  function a11yProps(index: number) {
    return {
      id: `simple-tab-${index}`,
      'aria-controls': `simple-tabpanel-${index}`,
    }
  }
  const onTabSelected = (
    _event: React.ChangeEvent<Record<string, never>>,
    newValue: string,
  ) => {
    setSelectedTab(newValue)
  }

  const onClickReplyButton = (e: React.MouseEvent<HTMLInputElement>) => {
    if (!messages || Object.keys(messages).length == 0) return
    const selectedMessageId = e.currentTarget.getAttribute('data-messageId')
    if (!selectedMessageId) return
    const msg = messages.get(Number(selectedMessageId))
    if (!msg) return
    setSelectedMessage(msg)
  }

  const onClickCloseButton = () => {
    setSelectedMessage(initialChatMessage)
  }

  const onSendQuestion = (data: MessageInputs) => {
    data.isQuestion = true
    onSendReply(data)
  }

  const onSendReply = (data: MessageInputs) => {
    if (!talk) return
    const api = new ChatMessageApi(
      new Configuration({ basePath: window.location.origin }),
    )
    api.apiV1ChatMessagesPost(
      CreateChatMessageRequest(
        data.chatMessage,
        talk.id,
        data.isQuestion,
        selectedMessage,
      ),
    )
    setSelectedMessage(initialChatMessage)
  }

  const onChecked = (
    _event: React.ChangeEvent<HTMLInputElement>,
    checked: boolean,
  ) => {
    setChecked(checked)
  }

  return (
    <Styled.Outer>
      <Styled.Container>
        <TabContext value={selectedTab}>
          <Styled.TabContainer
            value={selectedTab}
            onChange={onTabSelected}
            textColor="secondary"
            aria-label="simple tabs example"
          >
            <Styled.Tab label="Chat / QA" value="0" {...a11yProps(0)} />
            <Styled.Tab label="QA Only" value="1" {...a11yProps(1)} />
          </Styled.TabContainer>
          <Styled.TabPanel value="0">
            <ChatBox
              profile={profile}
              talk={talk}
              messages={messages}
              messageTypes={[
                ChatMessageMessageTypeEnum.Chat,
                ChatMessageMessageTypeEnum.Qa,
              ]}
              selectedMessage={selectedMessage}
              checked={checked}
              onClickReplyButton={onClickReplyButton}
              onSendReply={onSendReply}
              onClickCloseButton={onClickCloseButton}
            />
          </Styled.TabPanel>
          <Styled.TabPanel value="1">
            <ChatBox
              talk={talk}
              messages={messages}
              messageTypes={[ChatMessageMessageTypeEnum.Qa]}
              selectedMessage={selectedMessage}
              checked={checked}
              onClickReplyButton={onClickReplyButton}
              onClickCloseButton={onClickCloseButton}
              onSendReply={onSendReply}
            />
          </Styled.TabPanel>
        </TabContext>
        <ChatMessageForm
          isVisibleForm={isVisibleForm}
          selectedMessage={selectedMessage}
          onClickCloseButton={onClickCloseButton}
          onSendMessage={onSendReply}
          onSendQuestion={onSendQuestion}
          onCheck={onChecked}
          checked={checked}
        />
      </Styled.Container>
    </Styled.Outer>
  )
}
