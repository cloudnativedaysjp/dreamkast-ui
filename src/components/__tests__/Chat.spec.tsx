import React from 'react'
import renderer from 'react-test-renderer'
import { Talk } from '../../interfaces'
import Chat from '../chat'

const mockTalk: Talk = {
  id: 1,
  trackId: '1',
  vimeoId: '453943528',
  title: 'ほげふが',
  description: 'ぴよぴよ',
  speakers: ['jacopen1', 'r_takaishi2'],
}

test('Chat', () => {
  const component = renderer.create(<Chat talk={mockTalk} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
