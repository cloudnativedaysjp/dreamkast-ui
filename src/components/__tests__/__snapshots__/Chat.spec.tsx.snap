// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat 1`] = `
<div>
  <h2>
    Chat / QA
  </h2>
  <h3>
    ほげふが
  </h3>
  <ul
    className="MuiList-root MuiList-padding"
  >
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        foooo
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        bar
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        baz
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        hoge
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        fuga
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        piyo
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        fooo
      </span>
    </div>
  </ul>
</div>
`;
